### PR TITLE
Add support for min and max on boolean types.

### PR DESCRIPTION
**Pull request by DrTodd13**
Veros ocean simulator through Ramba generates min on booleans so adding support here.

### CODE REVIEW

**Comment by gmarkall** [comment]
Just had a quick look at this - doesn't it seem a little odd to squeeze the implementation for bools into the overloads for min and max on `IndexValue` types?

**Comment by DrTodd13** [comment]
> Just had a quick look at this - doesn't it seem a little odd to squeeze the implementation for bools into the overloads for min and max on `IndexValue` types?

I don't normally add these kind of overloads.  What is the standard approach in this case?  Should I just add a whole new function instead for each separate type?

**Comment by gmarkall** [comment]
> I don't normally add these kind of overloads. What is the standard approach in this case? Should I just add a whole new function instead for each separate type?

Yes - adding a new overload function for each separate type fits in well with the standard pattern for overloads (in my view / experience).

File: numba/cpython/builtins.py [hunk 1]
```diff
@@ -573,7 +573,11 @@ def min_impl(indval1, indval2):
                     return indval2
             return indval1
         return min_impl
-    elif isinstance(indval1, types.scalars.Boolean) and \
+
+
+@overload(min)
+def indval_min(indval1, indval2):
```
Comment (gmarkall) [inline numba/cpython/builtins.py:579]: I'd suggest naming the function and variables to reflect the fact they are for booleans, and not `IndexValue` pairs, e.g.
```suggestion
def bool_min(bool1, bool2):
```

etc.

**Review by gmarkall — DISMISSED** [review]
Many thanks for the fix! This looks good to me, but maybe someone else with a little more overload experience should give it a quick check over too.

Side note: this does highlight how it's a bit weird that our boolean values are not integers, unlike Python - e.g.:

```
In [2]: bool.mro()
Out[2]: [bool, int, object]
```

File: numba/tests/test_builtins.py [hunk 2]
```diff
@@ -838,6 +838,27 @@ def test_min_1_invalid_types_npm(self):
...
+
+        x_operands = [False, True]
+        y_operands = [False, True]
+        for x, y in itertools.product(x_operands, y_operands):
+            self.assertPreciseEqual(cfunc(x, y), pyfunc(x, y))
+
+    def test_max_bool1(self, flags=enable_pyobj_flags):
+        """
+        max(*args)
+        """
```
Comment (stuartarchibald) [inline numba/tests/test_builtins.py:853]:
```suggestion
        # tests max(<booleans>)
```
suggest this, else the doc strings end up in the test log which adds noise.

File: numba/tests/test_builtins.py [hunk 3]
```diff
@@ -838,6 +838,27 @@ def test_min_1_invalid_types_npm(self):
...
+        self.check_minmax_bool1(max_usecase1, flags)
+
+    def test_min_bool1(self, flags=enable_pyobj_flags):
+        """
+        min(*args)
+        """
```
Comment (stuartarchibald) [inline numba/tests/test_builtins.py:859]:
```suggestion
        # tests min(<booleans>)
```
as above.

File: numba/tests/test_builtins.py [hunk 4]
```diff
@@ -838,6 +838,27 @@ def test_min_1_invalid_types_npm(self):
         with self.assertTypingError():
             self.check_min_max_invalid_types(min_usecase1, flags=no_pyobj_flags)
 
+    def check_minmax_bool1(self, pyfunc, flags):
+        cr = compile_isolated(pyfunc, (types.bool_, types.bool_), flags=flags)
+        cfunc = cr.entry_point
+
+        x_operands = [False, True]
+        y_operands = [False, True]
+        for x, y in itertools.product(x_operands, y_operands):
```
Comment (stuartarchibald) [inline numba/tests/test_builtins.py:847]:
```suggestion
        operands = (False, True)
        for x, y in itertools.product(operands, operands):
```
`x_operands` and `y_operands` are the same, only need one definition?

File: numba/cpython/builtins.py [hunk 5]
```diff
@@ -575,6 +575,15 @@ def min_impl(indval1, indval2):
         return min_impl
 
 
+@overload(min)
+def boolval_min(val1, val2):
+    if isinstance(val1, types.scalars.Boolean) and \
+       isinstance(val2, types.scalars.Boolean):
```
Comment (stuartarchibald) [inline numba/cpython/builtins.py:581]:
```suggestion
    if isinstance(val1, types.Boolean) and \
       isinstance(val2, types.Boolean):
```
suggest using the top level re-export for consistency.

File: numba/cpython/builtins.py [hunk 6]
```diff
@@ -604,6 +613,15 @@ def max_impl(indval1, indval2):
         return max_impl
 
 
+@overload(max)
+def boolval_max(val1, val2):
+    if isinstance(val1, types.scalars.Boolean) and \
+       isinstance(val2, types.scalars.Boolean):
```
Comment (stuartarchibald) [inline numba/cpython/builtins.py:619]:
```suggestion
    if isinstance(val1, types.Boolean) and \
       isinstance(val2, types.Boolean):
```
suggest using the top level re-export for consistency.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @DrTodd13, a few minor things to resolve else looks good.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch!

**Pull request merged**